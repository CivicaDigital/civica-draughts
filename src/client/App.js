import React, { Component } from 'react';
import { DraughtsBoard } from './DraughtsBoard';
import { BoardController } from './board-controller';
import './app.css';

/**
 * Top level component for this application.
 * @demonstrates Wiring together of components, business logic and POJO state.
 * @potential Make player turn display more stylish. Display history and allow undo.
 * State could be moved to separate data store objects, with corresponding event subscription
 *{@link https://codeutopia.net/blog/2016/02/01/react-application-data-flow-where-and-how-to-store-your-data/}
 */
export class App extends Component {
  /**
  * @param {object} props The React properties.
  */
  constructor(props) {
    super(props);

    /** Stores the state of the board and enables user interactions. */
    this.boardController = new BoardController(props.size);

    /** Stores the state of this component. */
    this.state = {
      squares: this.boardController.startGame(props.fen),
      turn: this.boardController.turn
    };
  }

  /**
  * This is the pace to do any AJAX calls needed to initialise the component fully.
  * @method
  */
  componentDidMount() {

  }

  /**
  * Handles a click on a board square.
  * @method
  * @param {Square} square The square that was clicked.
  */
  handleSquareClick(square) {
    this.boardController.squareClicked(square);
    this.setState({ squares: this.boardController.board.squares, turn: this.boardController.turn });
  }

  /** undoButtonClick() {
    this.boardController.undoClicked();
  }
  */

  /**
  * Returns the React elements forming the page for which React will update the DOM accordingly.
  * @method
  */
  render() {
    const { squares, turn } = this.state;
    const turnText = turn && turn.blackTurn ? 'Black turn' : 'White turn';
    return (
      <div>
        <div>
          <h1>
            Draughts
          </h1>
          <div>
            { turnText }
          </div>
        </div>
        <div>
          <DraughtsBoard
            squares={squares}
            onClick={square => this.handleSquareClick(square)}
          />
        </div>
      </div>
    );
  }
}
