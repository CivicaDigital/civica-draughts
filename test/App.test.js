import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import App from '../src/client/App';

it('renders correctly when the board is empty', () => {
  const tree = renderer
    .create(<App />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders legal moves correctly when a square is clicked', () => {
  const wrapper = mount(<App />);
  expect(wrapper.find('button.selected')).toHaveLength(0);
  expect(wrapper.find('button.highlighted')).toHaveLength(0);

  wrapper.find('#square-27').simulate('click');
  wrapper.update();

  expect(wrapper.find('#square-27').hasClass('selected')).toEqual(true);
  expect(wrapper.find('button.selected')).toHaveLength(1);

  expect(wrapper.find('#square-23').hasClass('highlighted')).toEqual(true);
  expect(wrapper.find('#square-24').hasClass('highlighted')).toEqual(true);
  expect(wrapper.find('button.highlighted')).toHaveLength(2);
});
