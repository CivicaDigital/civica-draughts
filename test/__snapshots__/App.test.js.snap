// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when the board is empty 1`] = `
<div>
  <div>
    <h1 />
    <div>
      White turn
    </div>
  </div>
  <div>
    <div>
      <div>
        <button
          className="square unplayable"
          id="square--1"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-1"
          onClick={[Function]}
          type="button"
        >
          ⛂
        </button>
        <button
          className="square unplayable"
          id="square--2"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-2"
          onClick={[Function]}
          type="button"
        >
          ⛂
        </button>
        <button
          className="square unplayable"
          id="square--3"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-3"
          onClick={[Function]}
          type="button"
        >
          ⛂
        </button>
        <button
          className="square unplayable"
          id="square--4"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-4"
          onClick={[Function]}
          type="button"
        >
          ⛂
        </button>
      </div>
      <div>
        <button
          className="square playable"
          id="square-5"
          onClick={[Function]}
          type="button"
        >
          ⛂
        </button>
        <button
          className="square unplayable"
          id="square--5"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-6"
          onClick={[Function]}
          type="button"
        >
          ⛂
        </button>
        <button
          className="square unplayable"
          id="square--6"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-7"
          onClick={[Function]}
          type="button"
        >
          ⛂
        </button>
        <button
          className="square unplayable"
          id="square--7"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-8"
          onClick={[Function]}
          type="button"
        >
          ⛂
        </button>
        <button
          className="square unplayable"
          id="square--8"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
      </div>
      <div>
        <button
          className="square unplayable"
          id="square--9"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-9"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square unplayable"
          id="square--10"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-10"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square unplayable"
          id="square--11"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-11"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square unplayable"
          id="square--12"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-12"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
      </div>
      <div>
        <button
          className="square playable"
          id="square-13"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square unplayable"
          id="square--13"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-14"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square unplayable"
          id="square--14"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-15"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square unplayable"
          id="square--15"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-16"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square unplayable"
          id="square--16"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
      </div>
      <div>
        <button
          className="square unplayable"
          id="square--17"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-17"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square unplayable"
          id="square--18"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-18"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square unplayable"
          id="square--19"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-19"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square unplayable"
          id="square--20"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-20"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
      </div>
      <div>
        <button
          className="square playable"
          id="square-21"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square unplayable"
          id="square--21"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-22"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square unplayable"
          id="square--22"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-23"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square unplayable"
          id="square--23"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-24"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square unplayable"
          id="square--24"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
      </div>
      <div>
        <button
          className="square unplayable"
          id="square--25"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-25"
          onClick={[Function]}
          type="button"
        >
          ⛀
        </button>
        <button
          className="square unplayable"
          id="square--26"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-26"
          onClick={[Function]}
          type="button"
        >
          ⛀
        </button>
        <button
          className="square unplayable"
          id="square--27"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-27"
          onClick={[Function]}
          type="button"
        >
          ⛀
        </button>
        <button
          className="square unplayable"
          id="square--28"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-28"
          onClick={[Function]}
          type="button"
        >
          ⛀
        </button>
      </div>
      <div>
        <button
          className="square playable"
          id="square-29"
          onClick={[Function]}
          type="button"
        >
          ⛀
        </button>
        <button
          className="square unplayable"
          id="square--29"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-30"
          onClick={[Function]}
          type="button"
        >
          ⛀
        </button>
        <button
          className="square unplayable"
          id="square--30"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-31"
          onClick={[Function]}
          type="button"
        >
          ⛀
        </button>
        <button
          className="square unplayable"
          id="square--31"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
        <button
          className="square playable"
          id="square-32"
          onClick={[Function]}
          type="button"
        >
          ⛀
        </button>
        <button
          className="square unplayable"
          id="square--32"
          onClick={[Function]}
          type="button"
        >
          ‌ 
        </button>
      </div>
    </div>
  </div>
</div>
`;
